common/src/device/transport/mqtt/ClientWebSocketChannel.cs

```diff
@@ -0,0 +1,343 @@
+            _webSocket = webSocket;
+            _active = true;
+            Metadata = new ChannelMetadata(false, 16);
+            Configuration = new ClientWebSocketChannelConfig();
+            _writeCancellationTokenSource = new CancellationTokenSource();
+        }
+
+        public override IChannelConfiguration Configuration { get; }
+
+        public override bool Open => _webSocket.State == WebSocketState.Open;
+
+        public override bool Active => _active;
+
+        public override ChannelMetadata Metadata { get; }
+
+        public ClientWebSocketChannel Option<T>(ChannelOption<T> option, T value)
+        {
+            Contract.Requires(option != null);
+
+            Configuration.SetOption(option, value);
+            return this;
+        }
+
+        protected override EndPoint LocalAddressInternal { get; }
+
+        protected override EndPoint RemoteAddressInternal { get; }
+
+        protected override IChannelUnsafe NewUnsafe() => new WebSocketChannelUnsafe(this);
+
+        protected class WebSocketChannelUnsafe : AbstractUnsafe
+        {
+            public WebSocketChannelUnsafe(AbstractChannel channel)
+                : base(channel)
+            {
+            }
+
+            public override Task ConnectAsync(EndPoint remoteAddress, EndPoint localAddress)
+            {
+                throw new NotSupportedException("ClientWebSocketChannel does not support ConnectAsync()");
+            }
+
+            protected override void Flush0()
+            {
+                // Flush immediately only when there's no pending flush.
+                // If there's a pending flush operation, event loop will call FinishWrite() later,
+                // and thus there's no need to call it now.
+                if (((ClientWebSocketChannel)channel).WriteInProgress)
+                {
+                    return;
+                }
+
+                base.Flush0();
+            }
+        }
+
+        protected override bool IsCompatible(IEventLoop eventLoop) => true;
+
+        protected override void DoBind(EndPoint localAddress)
+        {
+            throw new NotSupportedException("ClientWebSocketChannel does not support DoBind()");
+        }
+
+        protected override void DoDisconnect()
+        {
+            throw new NotSupportedException("ClientWebSocketChannel does not support DoDisconnect()");
+        }
+
+        protected override async void DoClose()
+        {
+            try
+            {
+                WebSocketState webSocketState = _webSocket.State;
+                if (webSocketState != WebSocketState.Closed && webSocketState != WebSocketState.Aborted)
+                {
+                    // Cancel any pending write
+                    CancelPendingWrite();
+                    _active = false;
+
+                    using (var cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(30)))
+                    {
+                        await _webSocket.CloseAsync(
```

Comment:
if it acts the same as server version, it may deadlock. you can check our server websocket channel impl for details on how we work around this.

Reply:
You mean https://github.com/Microsoft/dotnet/issues/560?

Reply:
Yes. Good to see the fix is coming. Meanwhile it's likely a good idea not to deadlock executor's thread